internal/pkg/template/template.go

```diff
@@ -95,6 +100,21 @@ func New() *Template {
 	}
 }
 
+// AddonsArtifactPath prefixes the key with the addons artifact path.
+func AddonsArtifactPath(key string) string {
```

Comment:
I wonder if we should have a new package for all of these? so that it reads a bit better `template.MkdirSHA256` is not obvious that it's for s3.


Let me know what you think of this layout:
```go
// Package artifactpath holds functions to generate the S3 object path for artifacts.
package template/artifactpath

// Addons returns the path to store addon artifact files.
func Addons(key string) string

// CFNTemplate returns the path to store cloudformation templates.
func CFNTemplate(key string) string

// SHA256Func returns a function that will namespace a key with the sha256 hash of its content.
func SHA256Func(content []byte) func(string) string  {
    return func(key string) string {
        return path.Join(fmt.Sprintf("%x", sha256.Sum256(content)), key)
    }
}

// Chain applies the path functions sequentially to nest the key.
func Chain(key string, fn ...func(string) string) string 
```

This would be the client's experience:
```go
artifactpath.Chain(key, artifactpath.SHA256Func(content), artifactpath.CFNTemplate)
```

Reply:
Since we always want to add SHA256, I've inlined it into the function -> chaining was rendered unnecessary when SHA256 is inlined. Let me know what you think!